File: 2022/august/12-08-2022/lowest-common-ancestor-of-a-binary-search-tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode', prev=None) -> 'TreeNode':
        # return self.visit(root, root, p, q)
        if p.val == root.val or q.val == root.val:
            return prev
        if root.val < p.val and root.val < q.val:
            return self.lowestCommonAncestor(root.right, p, q, root)
        elif root.val > p.val and root.val > q.val:
            return self.lowestCommonAncestor(root.left, p, q, root)
        else:
            return root

File: 2022/august/12-08-2022/lowest-common-ancestor-of-a-binary-search-tree.rs
use std::rc::Rc;
use std::cell::RefCell;
type Node = Option<Rc<RefCell<TreeNode>>>;
impl Solution {
    pub fn lowest_common_ancestor(root: Node, p: Node, q: Node) -> Node {
        let (mut vp, mut vq) = (p.as_ref().unwrap().borrow().val, q.as_ref().unwrap().borrow().val);
        if vp>vq { vp+=vq; vq=vp-vq; vp-=vq; }
        match root {
            x if x==p => { p },
            x if x==q => { q },
            Some(node) => {
                let v = node.borrow().val;
                if vp<v && v<vq { return Some(node); }
                if vq<v { Solution::lowest_common_ancestor(node.borrow().left.clone(), p, q) }
                else { Solution::lowest_common_ancestor(node.borrow().right.clone(), p, q) }
            },
            None => None
        }
    }
}
